up-subscription/src/subscription_manager.rs

```diff
@@ -0,0 +1,746 @@
+#[derive(Debug)]
+pub(crate) enum SubscriptionEvent {
+    AddSubscription {
+        subscriber: SubscriberInfo,
+        topic: UUri,
+        respond_to: oneshot::Sender<SubscriptionStatus>,
+    },
+    RemoveSubscription {
+        subscriber: SubscriberInfo,
+        topic: UUri,
+        respond_to: oneshot::Sender<SubscriptionStatus>,
+    },
+    FetchSubscribers {
+        request: FetchSubscribersRequest,
+        respond_to: oneshot::Sender<FetchSubscribersResponse>,
+    },
+    FetchSubscriptions {
+        request: FetchSubscriptionsRequest,
+        respond_to: oneshot::Sender<FetchSubscriptionsResponse>,
+    },
+    // Purely for use during testing: get copy of current topic-subscriper ledger
+    #[cfg(test)]
+    GetTopicSubscribers {
+        respond_to: oneshot::Sender<HashMap<UUri, HashSet<SubscriberInfo>>>,
+    },
+    // Purely for use during testing: force-set new topic-subscriber ledger
+    #[cfg(test)]
+    SetTopicSubscribers {
+        topic_subscribers_replacement: HashMap<UUri, HashSet<SubscriberInfo>>,
+        respond_to: oneshot::Sender<()>,
+    },
+    // Purely for use during testing: get copy of current topic-subscriper ledger
+    #[cfg(test)]
+    GetRemoteTopics {
+        respond_to: oneshot::Sender<HashMap<UUri, State>>,
+    },
+    // Purely for use during testing: force-set new topic-subscriber ledger
+    #[cfg(test)]
+    SetRemoteTopics {
+        topic_subscribers_replacement: HashMap<UUri, State>,
+        respond_to: oneshot::Sender<()>,
+    },
+}
+
+// Internal subscription manager API - used to update on remote subscriptions (deal with _PENDING states)
+enum RemoteSubscriptionEvent {
+    RemoteSubscriptionStateUpdate { topic: UUri, state: State },
+}
+
+// Wrapper type, include all kinds of actions subscription manager knows
+enum Event {
+    LocalSubscription(SubscriptionEvent),
+    RemoteSubscription(RemoteSubscriptionEvent),
+}
+
+// Core business logic of subscription management - includes container data types for tracking subscriptions and remote subscriptions.
+// Interfacing with this purely works via channels, so we do not have to deal with mutexes and similar concepts.
+pub(crate) async fn handle_message(
+    own_uri: UUri,
+    up_client: Arc<dyn RpcClient>,
+    mut command_receiver: UnboundedReceiver<SubscriptionEvent>,
+    shutdown: Arc<Notify>,
+) {
+    helpers::init_once();
+
+    // track subscribers for topics - if you're in this list, you have SUBSCRIBED, otherwise you're considered UNSUBSCRIBED
+    #[allow(clippy::mutable_key_type)]
+    let mut topic_subscribers: HashMap<UUri, HashSet<SubscriberInfo>> = HashMap::new();
+
+    // for remote topics, we need to additionally deal with _PENDING states, this tracks states of these topics
+    #[allow(clippy::mutable_key_type)]
+    let mut remote_topics: HashMap<UUri, State> = HashMap::new();
```

Comment:
What happens if the uSubscription service falls over?

Should this be persisted even if this service crashes?

Reply:
Yeah, point taken - future improvement :-)